File: demo2.py
class Hajauttaja:
  def __init__(self, size):
    self.m = size
    self.t = [0] * self.m
  
  def h(self, k, i):
    return (k % self.m + i) % self.m
  
  def lisaa(self, k, v):
    for i in range(0, self.m):
      j = self.h(k, i)
      # 0 = vapaa
      # -1 = poistettu
      if self.t[j] == 0 or self.t[j] == -1:
        self.t[j] = {"avain": k, "arvo":v} 
        return True
    return False
  
  def etsi(self, k):
    for i in range(0, self.m):
      j = self.h(k, i)
      if self.t[0] == 0:
        return -1
      elif (self.t[j] != -1 and 
        type(self.t[j]) is dict and self.t[j]["avain"] == k):
        return j
    return -1
  
  def poista(self, k):
    j = self.etsi(k)
    if j < 0: # ei löytynyt
      return False
    else:
      temp = self.t[j]
      self.t[j] = -1
      return temp



haj = Hajauttaja(5)
haj.lisaa(19, 2)
haj.lisaa(88, 3)
haj.lisaa(8, 4)
print(haj.t)

print(haj.etsi(88))
print(haj.etsi(0))

print(haj.poista(88))
print(haj.poista(123456))
print(haj.t)

File: demo2.js
class Hajauttaja {
  constructor(size) {
    this.m = size;
    this.t = new Array(this.m);
    // 0 = vappa
    for (let i = 0; i < this.t.length; i++)
      this.t[i] = 0;
  }

  // käytetään lineaarista etsintää
  h(k, i) {
    return (k % this.m + i) % this.m;
  }

  lisaa(k, v) {
    for (let i = 0; i < this.m; i++) {
      const j = this.h(k, i);
      // 0 = vappa
      // -1 = poistettu
      if (this.t[j] === 0 || this.t[j] === -1) {
        this.t[j] = {"avain": k, "arvo":v};
        return true;
      }
    }
    return false;
  }

  etsi(k) {
    for (let i = 0; i < this.m; i++) {
      const j = this.h(k, i);
      // 0 = vappa
      // -1 = poistettu
      if (this.t[j] === 0)
        return -1;
      else if ( this.t[j] != -1 && this.t[j].avain === k)
        return j; // palauttaa indexin
    }
    return -1;
  }

  poista(k) {
    const j = this.etsi(k);
    if (j < 0) // ei löytynyt
      return false;
    else {
      const temp = this.t[j];
      this.t[j] = -1;
      return temp; 
    }
  }
}

let haj = new Hajauttaja(5);

console.log(haj.t);
haj.lisaa(666, 11);
haj.lisaa(0, 55);
haj.lisaa(12345, 99);
console.log(haj.t);

console.log(haj.etsi(666));
console.log(haj.etsi(6));

console.log(haj.poista(666))
console.log(haj.t);

console.log(haj.poista(12345))
console.log(haj.poista(12345))
console.log(haj.t);